### DISCUSSION TITLE
Apple Macbook Air touchpad device name?

### DISCUSSION THREAD

**Discussion by guttermonk** [bulletmark/libinput-gestures#366] [Q&A]
When I run `libinput list-devices` I get the following two devices:

```
Device:           Apple Inc. Apple Internal Keyboard / Trackpad
Kernel:           /dev/input/event1
Group:            7
Seat:             seat0, default
Capabilities:     keyboard
Tap-to-click:     n/a
Tap-and-drag:     n/a
Tap drag lock:    n/a
Left-handed:      n/a
Nat.scrolling:    n/a
Middle emulation: n/a
Calibration:      n/a
Scroll methods:   none
Click methods:    none
Disable-w-typing: n/a
Disable-w-trackpointing: n/a
Accel profiles:   n/a
Rotation:         0.0
  
Device:           bcm5974
Kernel:           /dev/input/event10
Group:            7
Seat:             seat0, default
Size:             105x75mm
Capabilities:     pointer gesture
Tap-to-click:     disabled
Tap-and-drag:     enabled
Tap drag lock:    disabled
Left-handed:      disabled
Nat.scrolling:    enabled
Middle emulation: disabled
Calibration:      n/a
Scroll methods:   *two-finger edge
Click methods:    button-areas *clickfinger
Disable-w-typing: enabled
Disable-w-trackpointing: enabled
Accel profiles:   flat *adaptive custom
Rotation:         n/a
```

I tried adding `device bcm5974` to my config, which worked fine at first but, after hibernating, gestures stopped working. So I tried changing it to `device Apple Inc. Apple Internal Keyboard / Trackpad` but that killed gestures all together. When I set it to `device all` it works fine and survives hibernation, but as I understand it, setting it to `device all` isn't ideal. Any suggestions?

**Comment by bulletmark** [comment]
Please read the [TROUBLE-SHOOTING](https://github.com/bulletmark/libinput-gestures#troubleshooting) section, particularly the opening paragraph.

**Comment by guttermonk** [comment]
Thank you for getting back to me. I read through the trouble-shooting section but it wasn't much help. Looking at the raw data for `device bcm5974`, gestures are received normally but as soon as I hibernate it stops receiving anything.

I also read through the following issues on the same topic: [68](https://github.com/bulletmark/libinput-gestures/issues/68), [160](https://github.com/bulletmark/libinput-gestures/issues/160), [177](https://github.com/bulletmark/libinput-gestures/issues/177), [194](https://github.com/bulletmark/libinput-gestures/issues/194), and [228](https://github.com/bulletmark/libinput-gestures/issues/228). Looks like most people are either using [a restart script for Gnome](https://github.com/bulletmark/libinput-gestures/issues/163#issuecomment-428340925) and [KDE](https://github.com/bulletmark/libinput-gestures/issues/194#issuecomment-457662356), or they end up giving up and using `device all`. Since I'm using NixOS, unfortunately neither of those scripts work on my system.

Here's the [output of my libinput-gestures -l](https://github.com/bulletmark/libinput-gestures/issues/194#issuecomment-2305995799).

Here's this too:
```
ls -la /dev/input/by-path/
drwxr-xr-x - root 25 Aug 06:33 .
drwxr-xr-x - root 25 Aug 06:33 ..
lrwxrwxrwx - root 25 Aug 06:17 pci-0000:00:14.0-usb-0:1:1.0-event-kbd -> ../event0
lrwxrwxrwx - root 25 Aug 06:16 pci-0000:00:14.0-usb-0:5:1.1-event-kbd -> ../event1
lrwxrwxrwx - root 25 Aug 06:33 pci-0000:00:14.0-usb-0:5:1.2-event-mouse -> ../event2
lrwxrwxrwx - root 25 Aug 06:33 pci-0000:00:14.0-usb-0:5:1.2-mouse -> ../mouse0
lrwxrwxrwx - root 25 Aug 06:17 pci-0000:00:14.0-usbv2-0:1:1.0-event-kbd -> ../event0
lrwxrwxrwx - root 25 Aug 06:16 pci-0000:00:14.0-usbv2-0:5:1.1-event-kbd -> ../event1
lrwxrwxrwx - root 25 Aug 06:33 pci-0000:00:14.0-usbv2-0:5:1.2-event-mouse -> ../event2
lrwxrwxrwx - root 25 Aug 06:33 pci-0000:00:14.0-usbv2-0:5:1.2-mouse -> ../mouse0

ls -la /dev/input/by-id/
drwxr-xr-x - root 25 Aug 06:33 .
drwxr-xr-x - root 25 Aug 06:33 ..
lrwxrwxrwx - root 25 Aug 06:16 usb-Apple_Inc._Apple_Internal_Keyboard___Trackpad_DQ63363GAVRF94RATBE-if01-event-kbd -> ../event1
lrwxrwxrwx - root 25 Aug 06:33 usb-Apple_Inc._Apple_Internal_Keyboard___Trackpad_DQ63363GAVRF94RATBE-if02-event-mouse -> ../event2
lrwxrwxrwx - root 25 Aug 06:33 usb-Apple_Inc._Apple_Internal_Keyboard___Trackpad_DQ63363GAVRF94RATBE-if02-mouse -> ../mouse0
lrwxrwxrwx - root 25 Aug 06:17 usb-Yubico_YubiKey_OTP+FIDO+CCID-event-kbd -> ../event0
```

**Comment by guttermonk** [comment]
Today I tried setting each of the /dev/input/ paths listed above, and got the following results.  

Worked at first but didn't survive hibernation:  
```
device bcm5974  
device /dev/input/by-path/pci-0000:00:14.0-usb-0:5:1.2-event-mouse
device /dev/input/by-path/pci-0000:00:14.0-usbv2-0:5:1.2-event-mouse  
device /dev/input/by-path/pci-0000:00:14.0-usbv2-0:5:1.2-mouse  
device /dev/input/by-id/usb-Apple_Inc._Apple_Internal_Keyboard___Trackpad_DQ63363GAVRF94RATBE-if02-event-mouse  
```
  
Did not work at all:  
```
device Apple Inc. Apple Internal Keyboard / Trackpad  
device /dev/input/by-path/pci-0000:00:14.0-usb-0:5:1.2-mouse  
device /dev/input/by-id/usb-Apple_Inc._Apple_Internal_Keyboard___Trackpad_DQ63363GAVRF94RATBE-if02-mouse
```

**Comment by bulletmark** [comment]
From your output of `libinput-gestures -l`, I don't recognise the version of `libinput-gestures` you are running. How did you install it? Can you please install and try the latest version?

**Comment by guttermonk** [comment]
I've installed it from the [Nix package repo](https://search.nixos.org/packages?channel=24.05&show=libinput-gestures&from=0&size=50&sort=relevance&type=packages&query=libinput-gestures). They currently have version 2.76, which is the same version that's in the AUR and it is the same version that I'm currently running.

**Comment by bulletmark** [comment]
I didn't recognize that hash because the code is different. Nix patches the code to change some things related to their environment although it seems the changes are benign.

**Comment by bulletmark** [comment]
I've seen others report problems with suspend where, e.g. the device hangs or the device name changes, and for that I created another [dbus-action](https://github.com/bulletmark/dbus-action/tree/master) utility which is mentioned in the [README here](https://github.com/bulletmark/libinput-gestures?tab=readme-ov-file#automatic-stoprestart-on-d-bus-events-such-as-suspend). It works by restarting libinput-gestures (i.e. `libinput-gestures-setup restart`) after system resume so that it rediscovers where the touchpad is. However, I don't think you can use that on a Mac which I don't think offers D-Bus. You can probably find some other tool that can do something like that though.

**Comment by guttermonk** [comment]
I'll have to look into the possibility of using D-Bus in NixOS on a Macbook. For some reason, the `libinput-gestures-setup restart` command isn't working at the moment on my NixOS (it worked fine when I used Arch), but it may be due to some configuration missing from my configuration.nix file. I'll investigate more with the nix community.

In the meantime, do you happen to know if a bug has been submitted for the libinput package? Based on [your post on 194](https://github.com/bulletmark/libinput-gestures/issues/194#issuecomment-457547823), do you still think this is an issue with `libinput debug-events` not automatically resuming after a suspend for some users/touchpads? Peter Huterrer was suggesting in [this post](https://gitlab.freedesktop.org/libinput/libinput/-/issues/374#note_257693) that a work around would be for libinput-gestures to use the udev backend, if I'm not mistaken. I don't know a lot about these systems, so I'm not sure if that's possible. Just curious what your thoughts are.

**Comment by guttermonk** [comment]
On NixOS running Hyprland, I found that I can issue the following commands using either [hypridle](https://wiki.hyprland.org/Hypr-Ecosystem/hypridle/) or [swayidle](https://github.com/swaywm/swayidle/blob/master/swayidle.1.scd), before hibernation: 
`pkill -f gestures`

And after hibernation:
`libinput-gestures`

This solves the issue without having to use the `device all` method. I hope this helps anyone that stumbles on this in the future.

**Reply by bulletmark** [reply]
A `pkill` is crude. If you want to stop `libinput-gestures` then do a `libinput-gestures-setup stop`. If you want to start or restart it do a `libinput-gestures-setup start` or  `libinput-gestures-setup restart` .

**Reply by guttermonk** [reply]
I started [a thread on the Nix forum](https://discourse.nixos.org/t/missing-libinput-gestures-starting-stopping-commands/51545) to get to the bottom of that.

**Reply by guttermonk** [reply]
I was able to get some help to get libinput-gestures' systemd commands working on NixOS. I added `systemctl --user start libinput-gestures &` to my Hyprland startup script, and the following to my hyprland.conf:

`exec-once = swayidle -w after-resume 'systemctl --user start libinput-gestures' before-sleep 'systemctl --user stop libinput-gestures'`

Everything appears to be working correctly. Gestures are coming back after resuming from hibernation. Any down-sides to doing it like this?

**Reply by bulletmark** [reply]
Assume you are asking me but sorry I am not familiar at all about NixOS, Hyprland, or swayidle. It looks sensible though.
